File: examples/go/handlers/file_io.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ── go_file_write ─────────────────────────────────────────────────────────────

type fileWritePayload struct {
	Filename string `json:"filename"` // written under examples/output/
	Content  string `json:"content"`
	Append   bool   `json:"append"` // default false (overwrite)
}

type fileWriteResult struct {
	Path         string `json:"path"`
	BytesWritten int    `json:"bytes_written"`
	Mode         string `json:"mode"` // "write" or "append"
}

// FileWrite writes (or appends) content to a file under examples/output/.
//
// Payload: {"filename": "hello.txt", "content": "Hello!\n", "append": false}
// Output:  examples/output/<filename>
func FileWrite(ctx context.Context, payload []byte) ([]byte, error) {
	var p fileWritePayload
	if err := json.Unmarshal(payload, &p); err != nil {
		return nil, fmt.Errorf("invalid payload: %w", err)
	}
	if p.Filename == "" {
		return nil, fmt.Errorf("filename is required")
	}
	// Sanitise: no path traversal
	if strings.Contains(p.Filename, "..") || filepath.IsAbs(p.Filename) {
		return nil, fmt.Errorf("filename must be a relative name without ..")
	}

	outPath := filepath.Join(outputDir(), p.Filename)
	if err := os.MkdirAll(filepath.Dir(outPath), 0o755); err != nil {
		return nil, fmt.Errorf("mkdir: %w", err)
	}

	flag := os.O_CREATE | os.O_WRONLY
	if p.Append {
		flag |= os.O_APPEND
	} else {
		flag |= os.O_TRUNC
	}

	f, err := os.OpenFile(outPath, flag, 0o644)
	if err != nil {
		return nil, fmt.Errorf("open %s: %w", outPath, err)
	}
	defer f.Close()

	n, err := f.WriteString(p.Content)
	if err != nil {
		return nil, fmt.Errorf("write: %w", err)
	}

	mode := "write"
	if p.Append {
		mode = "append"
	}

	return json.Marshal(fileWriteResult{
		Path:         outPath,
		BytesWritten: n,
		Mode:         mode,
	})
}

// ── go_file_read ──────────────────────────────────────────────────────────────

type fileReadPayload struct {
	Path     string `json:"path"`      // path to read (relative to cwd)
	MaxBytes int    `json:"max_bytes"` // default 65536
}

type fileReadResult struct {
	Path       string `json:"path"`
	SizeBytes  int64  `json:"size_bytes"`
	Content    string `json:"content"`
	Truncated  bool   `json:"truncated"`
	OutputFile string `json:"output_file"` // log copy written here
}

// FileRead reads a file and writes a log copy to examples/output/.
//
// Payload: {"path": "examples/output/hello.txt"}
// Output:  examples/output/go_file_read_<timestamp>.txt
func FileRead(_ context.Context, payload []byte) ([]byte, error) {
	var p fileReadPayload
	if err := json.Unmarshal(payload, &p); err != nil {
		return nil, fmt.Errorf("invalid payload: %w", err)
	}
	if p.Path == "" {
		return nil, fmt.Errorf("path is required")
	}
	if p.MaxBytes <= 0 {
		p.MaxBytes = 64 * 1024
	}

	info, err := os.Stat(p.Path)
	if err != nil {
		return nil, fmt.Errorf("stat %s: %w", p.Path, err)
	}

	data, err := os.ReadFile(p.Path)
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", p.Path, err)
	}

	truncated := false
	content := data
	if len(content) > p.MaxBytes {
		content = content[:p.MaxBytes]
		truncated = true
	}

	logContent := fmt.Sprintf(
		"=== go_file_read log ===\npath:      %s\nsize:      %d bytes\nread_at:   %s\ntruncated: %v\n\n--- content ---\n%s",
		p.Path, info.Size(), time.Now().Format(time.RFC3339), truncated, content,
	)
	outPath, err := writeOutputText(
		fmt.Sprintf("go_file_read_%d.txt", time.Now().UnixNano()),
		logContent,
	)
	if err != nil {
		return nil, err
	}

	return json.Marshal(fileReadResult{
		Path:       p.Path,
		SizeBytes:  info.Size(),
		Content:    string(content),
		Truncated:  truncated,
		OutputFile: outPath,
	})
}

File: examples/python/handlers/file_io.py
"""py_file_write / py_file_read — simple file I/O handlers."""
from __future__ import annotations

import logging
import uuid
from pathlib import Path

OUTPUT_DIR = Path(__file__).parent.parent.parent / "output"
logger = logging.getLogger(__name__)


def handle_file_write(payload: dict) -> dict:
    """
    Write (or append) text to a file under examples/output/.

    Payload: {"filename": "hello.txt", "content": "Hello!\\n", "append": false}
    Output:  examples/output/<filename>
    """
    filename = payload.get("filename")
    if not filename:
        raise ValueError("filename is required")
    content = payload.get("content", "")
    append = payload.get("append", False)

    # Safety: no path traversal
    out_path = (OUTPUT_DIR / filename).resolve()
    if not str(out_path).startswith(str(OUTPUT_DIR.resolve())):
        raise ValueError("filename must stay within examples/output/")

    out_path.parent.mkdir(parents=True, exist_ok=True)
    mode = "a" if append else "w"
    out_path.open(mode, encoding="utf-8").write(content)

    bytes_written = len(content.encode())
    logger.info("wrote %d bytes → %s (mode=%s)", bytes_written, out_path, mode)
    return {"path": str(out_path), "bytes_written": bytes_written, "mode": mode}


def handle_file_read(payload: dict) -> dict:
    """
    Read a file and write a log copy to examples/output/.

    Payload: {"path": "examples/output/hello.txt", "max_bytes": 65536}
    """
    path_str = payload.get("path")
    if not path_str:
        raise ValueError("path is required")
    max_bytes = payload.get("max_bytes", 64 * 1024)

    file_path = Path(path_str)
    if not file_path.exists():
        raise FileNotFoundError(f"file not found: {file_path}")

    raw = file_path.read_bytes()
    truncated = len(raw) > max_bytes
    content = raw[:max_bytes].decode("utf-8", errors="replace")

    log_content = (
        f"=== py_file_read log ===\n"
        f"path:      {file_path}\n"
        f"size:      {len(raw)} bytes\n"
        f"truncated: {truncated}\n\n"
        f"--- content ---\n{content}"
    )

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    out_path = OUTPUT_DIR / f"py_file_read_{uuid.uuid4().hex[:8]}.txt"
    out_path.write_text(log_content, encoding="utf-8")

    logger.info("read %d bytes from %s → log at %s", len(raw), file_path, out_path)
    return {
        "path": str(file_path),
        "size_bytes": len(raw),
        "truncated": truncated,
        "output_file": str(out_path),
    }
